internal/util/image/pull.go

```diff
@@ -0,0 +1,265 @@
+package image
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"iter"
+	"os"
+	"time"
+
+	"github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/image"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/pkg/blobinfocache/none"
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+	"github.com/go-logr/logr"
+	ocispecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Puller interface {
+	Pull(context.Context, string, string, Cache) (fs.FS, reference.Canonical, time.Time, error)
+}
+
+type LayerData struct {
+	Reader io.Reader
+	Index  int
+	Err    error
+}
+
+type Cache interface {
+	Fetch(context.Context, string, reference.Canonical) (fs.FS, time.Time, error)
+	Store(context.Context, string, reference.Named, reference.Canonical, ocispecv1.Image, iter.Seq[LayerData]) (fs.FS, time.Time, error)
+	DeleteID(context.Context, string) error
+	GarbageCollect(context.Context, string, reference.Canonical) error
+}
+
+var insecurePolicy = []byte(`{"default":[{"type":"insecureAcceptAnything"}]}`)
+
+type ContainersImagePuller struct {
+	SourceCtxFunc func(context.Context) (*types.SystemContext, error)
+}
+
+func (p *ContainersImagePuller) Pull(ctx context.Context, id string, ref string, cache Cache) (fs.FS, reference.Canonical, time.Time, error) {
```

Comment:
Should probably rename `id` to `owner` everywhere it gets passed around. If two different owners request the same image, we'll end up pulling the image separately. This is by design so that if/when we introduce more configuration in our APIs around image pulling, we can make sure that each ClusterExtension is sandboxed from the other and is unable to re-use an image that another ClusterExtension pulled. This is especially important, for example, if two ClusterExtensions ask for the same private image, but only one ClusterExtension has permission to pull it.


```suggestion
func (p *ContainersImagePuller) Pull(ctx context.Context, owner string, ref string, cache Cache) (fs.FS, reference.Canonical, time.Time, error) {
```